documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md

```diff
@@ -0,0 +1,277 @@
+| `"@nuxtjs/redirect-module"`        | Define Nitro routeRules in `nuxt.config`                                                             | https://nitro.unjs.io/config/#routerules                                                                                                                 |
+| `"@nuxtjs/sentry"`                 | No official support                                                                                  | Workaround in https://github.com/nuxt-community/sentry-module/issues/530                                                                                 |
+| `"vue-plausible"`                  | Update to current version that supports Nuxt 3 or https://github.com/nuxt-modules/plausible          | Note: enabling outbound tracking breaks links with target blank in `vue-plausible`                                                                       |
+| `"@nuxtjs/sitemap"`                | Replace with `nuxt-simple-sitemap`                                                                   | https://github.com/harlan-zw/nuxt-simple-sitemap                                                                                                         |
+
+We should probably add [`nuxt-vitest`](https://github.com/danielroe/nuxt-vitest)
+for testing in Nuxt. The package is under active development, so we should pin
+to the patch version
+
+The migration to Nuxt 3 also means migration from Vue 2 (with added composition
+API) to Vue 3. Below is a list of breaking changes that need to be addressed in
+Openverse.
+
+#### Vue 3 breaking changes
+
+- [ ] `v-model`
+      [changes](https://v3-migration.vuejs.org/breaking-changes/v-model.html)
+- [ ] `key` usage on `template v-for` instead of the components inside that
+      `template`
+- [ ] [`v-on:event.native` removed](https://v3-migration.vuejs.org/breaking-changes/v-on-native-modifier-removed.html) -
+      after removing `.native`, check that links work as expected, sending the
+      analytics events
+- [ ] [`$listeners` has been removed / merged into `$attrs`](https://v3-migration.vuejs.org/breaking-changes/listeners-removed)-
+      remove the `$listeners` everywhere and make sure that we have
+      `v-bind="$attrs"` everywhere we had `$listeners` before
+- [ ] [`$attrs` includes `class` & `style`](https://v3-migration.vuejs.org/breaking-changes/attrs-includes-class-style.html) -
+      remove `v-bind="$attrs"` from the root element in components where
+      `inheritAttrs="false"`
+- [ ] `Vue.set` is not necessary anymore to keep array/object reactivity
+- [ ] [async components change of syntax](https://v3-migration.vuejs.org/breaking-changes/async-components.html)
+      (for components that use `Comp = () => import("component.vue")`)
+- [ ] [`emits` option](https://v3-migration.vuejs.org/breaking-changes/emits-option.html) -
+      replace the `defineEvents` custom implementation
```

Comment:
These changes do not replace `defineEvent`, as far as I can tell. The `emits` feature described in the link is actually required for `defineEvent` to work in the first place, the docstring for the function specifies that it's a workaround for TS in Composition API: https://github.com//WordPress/openverse/blob/697f62f01a32cb7fcf2f4a7627650a113cba40da/frontend/src/types/emits.ts#L9-L18

I'm surprised that Vue doesn't have a built-in solution for this though, it's very silly to need to write an empty validation function (maybe one that even gets included in your final bundle, I'm not sure if Nuxt shakes those out in build) just to get types for events.

In any case, the `defineEvent` utility was already developed for the feature linked on the documentation page and I don't see anything there or on the Vue 3 `emits` documentation for ways to define the event types without the strange empty validator function workaround.

Reply:
I was actually thinking of `defineEmits`: https://vuejs.org/api/sfc-script-setup.html#defineprops-defineemits, more on typing emits: https://vuejs.org/guide/typescript/composition-api.html#typing-component-emits

However, they are only available in components that use `script setup`. 

I think the best course here would be to keep using `defineEvent` with `emits`, and refactor the components to `script setup` and `defineEmits` later, after this project is done. It could be a separate milestone for converting all of the components that emit events to `script setup`.

Reply:
> I think the best course here would be to keep using defineEvent with emits, and refactor the components to script setup and defineEmits later, after this project is done. It could be a separate milestone for converting all of the components that emit events to script setup.

Sounds good to me! I'm now remembering that I wrote `defineEvent` in frustration that `defineEmits` didn't work in traditional components :sweat_smile: